### ISSUE TITLE
InstabugActivityLifecycleListener violates android.os.strictmode.DiskReadViolation

### ISSUE THREAD

**Issue by stari4ek** [Instabug/Instabug-Android#268]
### Steps to Reproduce the Problem

1. Enable StrictMode checks (https://developer.android.com/reference/android/os/StrictMode) for the app
2. Start app with Instabug integrated
3. Activity should be started
4. `DiskReadViolation` is logged/fired about Disk access from the main thread

### Expected Behavior

No disk I/O from the main/UI thread

### Actual Behavior

I/O disk operations from the main thread which slows down App's UI creation and violated Strict mode policy
```
2020-01-09 18:15:45.501 8113-8113/by.stari4ek.tvirl.debug D/StrictMode: StrictMode policy violation; ~duration=194 ms: android.os.strictmode.DiskReadViolation
        at android.os.StrictMode$AndroidBlockGuardPolicy.onReadFromDisk(StrictMode.java:1504)
        at java.io.UnixFileSystem.getSpace(UnixFileSystem.java:361)
        at java.io.File.getUsableSpace(File.java:1868)
        at com.instabug.library.util.DeviceStateProvider.getFreeStorage(DeviceStateProvider.java:2)
        at com.instabug.library.tracking.a.onActivityResumed(InstabugActivityLifecycleListener.java:5)
        at android.app.Application.dispatchActivityResumed(Application.java:239)
        at android.app.Activity.onResume(Activity.java:1343)
        at androidx.fragment.app.FragmentActivity.onResume(FragmentActivity.java:456)
        at by.stari4ek.iptv4atv.tvinput.ui.main.MainMenuActivity.onResume(MainMenuActivity.java:48)
        at android.app.Instrumentation.callActivityOnResume(Instrumentation.java:1412)
        at android.app.Activity.performResume(Activity.java:7292)
        at android.app.ActivityThread.performResumeActivity(ActivityThread.java:3776)
        at android.app.ActivityThread.handleResumeActivity(ActivityThread.java:3816)
        at android.app.servertransaction.ResumeActivityItem.execute(ResumeActivityItem.java:51)
        at android.app.servertransaction.TransactionExecutor.executeLifecycleState(TransactionExecutor.java:145)
        at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:70)
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1808)
        at android.os.Handler.dispatchMessage(Handler.java:106)
        at android.os.Looper.loop(Looper.java:193)
        at android.app.ActivityThread.main(ActivityThread.java:6669)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:493)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:858)
```

### Instabug integration code
*Make sure to remove your app token.*

### SDK Version

9.0.4

### Android Version

tested with API 28

### Device Model

Emulator

**Comment by stari4ek** [comment]
btw, is there any way to fully disable UI tracking?

**Comment by stari4ek** [comment]
And there're bunch of other:
```
D/StrictMode: StrictMode policy violation; ~duration=127 ms: android.os.strictmode.DiskReadViolation
        at android.os.StrictMode$AndroidBlockGuardPolicy.onReadFromDisk(StrictMode.java:1504)
        at libcore.io.BlockGuardOs.open(BlockGuardOs.java:206)
        at libcore.io.IoUtils.canOpenReadOnly(IoUtils.java:163)
        at dalvik.system.DexPathList$NativeLibraryElement.findNativeLibrary(DexPathList.java:833)
        at dalvik.system.DexPathList.findLibrary(DexPathList.java:548)
        at dalvik.system.BaseDexClassLoader.findLibrary(BaseDexClassLoader.java:179)
        at java.lang.Runtime.loadLibrary0(Runtime.java:1005)
        at java.lang.System.loadLibrary(System.java:1669)
        at com.instabug.library.network.service.synclogs.SyncLogKeyProvider.<clinit>(SyncLogKeyProvider.java:1)
        at com.instabug.library.network.service.synclogs.b.<init>(SyncLogFacade.java:8)
        at com.instabug.library.network.service.synclogs.b.c(SyncLogFacade.java:2)
        at com.instabug.library.b.x(InstabugDelegate.java:1)
        at com.instabug.library.b.y(InstabugDelegate.java:9)
        at com.instabug.library.b.h(InstabugDelegate.java:1)
        at com.instabug.library.b$a.a(InstabugDelegate.java:19)
        at com.instabug.library.b$a.accept(InstabugDelegate.java:1)
        at io.reactivex.internal.observers.LambdaObserver.onNext(LambdaObserver.java:63)
        at io.reactivex.subjects.PublishSubject$PublishDisposable.onNext(PublishSubject.java:308)
        at io.reactivex.subjects.PublishSubject.onNext(PublishSubject.java:228)
        at com.instabug.library.core.eventbus.EventBus.post(EventBus.java:1)
        at com.instabug.library.g.a(SessionManager.java:14)
        at com.instabug.library.g.e(SessionManager.java:3)
        at com.instabug.library.tracking.a.onActivityResumed(InstabugActivityLifecycleListener.java:15)
        at android.app.Application.dispatchActivityResumed(Application.java:239)
        at android.app.Activity.onResume(Activity.java:1343)
```

```
D/StrictMode: StrictMode policy violation; ~duration=111 ms: android.os.strictmode.DiskReadViolation
        at android.os.StrictMode$AndroidBlockGuardPolicy.onReadFromDisk(StrictMode.java:1504)
        at android.database.sqlite.SQLiteConnection.applyBlockGuardPolicy(SQLiteConnection.java:1055)
        at android.database.sqlite.SQLiteConnection.executeForCursorWindow(SQLiteConnection.java:856)
        at android.database.sqlite.SQLiteSession.executeForCursorWindow(SQLiteSession.java:836)
        at android.database.sqlite.SQLiteQuery.fillWindow(SQLiteQuery.java:62)
        at android.database.sqlite.SQLiteCursor.fillWindow(SQLiteCursor.java:149)
        at android.database.sqlite.SQLiteCursor.getCount(SQLiteCursor.java:137)
        at android.database.AbstractCursor.moveToPosition(AbstractCursor.java:220)
        at android.database.AbstractCursor.moveToFirst(AbstractCursor.java:259)
        at com.instabug.library.internal.storage.cache.user.a.a(UserDbHelper.java:70)
        at com.instabug.library.internal.storage.cache.user.UserCacheManager.insertIfNotExists(UserCacheManager.java:2)
        at com.instabug.library.user.b.g(UserManager.java:20)
        at com.instabug.library.b.x(InstabugDelegate.java:2)
        at com.instabug.library.b.y(InstabugDelegate.java:9)
        at com.instabug.library.b.h(InstabugDelegate.java:1)
        at com.instabug.library.b$a.a(InstabugDelegate.java:19)
        at com.instabug.library.b$a.accept(InstabugDelegate.java:1)
        at io.reactivex.internal.observers.LambdaObserver.onNext(LambdaObserver.java:63)
        at io.reactivex.subjects.PublishSubject$PublishDisposable.onNext(PublishSubject.java:308)
        at io.reactivex.subjects.PublishSubject.onNext(PublishSubject.java:228)
        at com.instabug.library.core.eventbus.EventBus.post(EventBus.java:1)
        at com.instabug.library.g.a(SessionManager.java:14)
        at com.instabug.library.g.e(SessionManager.java:3)
        at com.instabug.library.tracking.a.onActivityResumed(InstabugActivityLifecycleListener.java:15)
        at android.app.Application.dispatchActivityResumed(Application.java:239)
        at android.app.Activity.onResume(Activity.java:1343)
        at androidx.fragment.app.FragmentActivity.onResume(FragmentActivity.java:456)
```
```
D/StrictMode: StrictMode policy violation; ~duration=101 ms: android.os.strictmode.DiskReadViolation
        at android.os.StrictMode$AndroidBlockGuardPolicy.onReadFromDisk(StrictMode.java:1504)
        at android.database.sqlite.SQLiteConnection.applyBlockGuardPolicy(SQLiteConnection.java:1055)
        at android.database.sqlite.SQLiteConnection.executeForCursorWindow(SQLiteConnection.java:856)
        at android.database.sqlite.SQLiteSession.executeForCursorWindow(SQLiteSession.java:836)
        at android.database.sqlite.SQLiteQuery.fillWindow(SQLiteQuery.java:62)
        at android.database.sqlite.SQLiteCursor.fillWindow(SQLiteCursor.java:149)
        at android.database.sqlite.SQLiteCursor.getCount(SQLiteCursor.java:137)
        at android.database.AbstractCursor.moveToPosition(AbstractCursor.java:220)
        at android.database.AbstractCursor.moveToFirst(AbstractCursor.java:259)
        at com.instabug.library.internal.storage.cache.user.a.a(UserDbHelper.java:70)
        at com.instabug.library.internal.storage.cache.user.UserCacheManager.getLastSeen(UserCacheManager.java:1)
        at com.instabug.library.internal.storage.cache.user.UserCacheManager.insertUser(UserCacheManager.java:1)
        at com.instabug.library.internal.storage.cache.user.UserCacheManager.insertIfNotExists(UserCacheManager.java:3)
        at com.instabug.library.user.b.g(UserManager.java:20)
        at com.instabug.library.b.x(InstabugDelegate.java:2)
        at com.instabug.library.b.y(InstabugDelegate.java:9)
        at com.instabug.library.b.h(InstabugDelegate.java:1)
        at com.instabug.library.b$a.a(InstabugDelegate.java:19)
        at com.instabug.library.b$a.accept(InstabugDelegate.java:1)
        at io.reactivex.internal.observers.LambdaObserver.onNext(LambdaObserver.java:63)
        at io.reactivex.subjects.PublishSubject$PublishDisposable.onNext(PublishSubject.java:308)
        at io.reactivex.subjects.PublishSubject.onNext(PublishSubject.java:228)
        at com.instabug.library.core.eventbus.EventBus.post(EventBus.java:1)
        at com.instabug.library.g.a(SessionManager.java:14)
        at com.instabug.library.g.e(SessionManager.java:3)
        at com.instabug.library.tracking.a.onActivityResumed(InstabugActivityLifecycleListener.java:15)
        at android.app.Application.dispatchActivityResumed(Application.java:239)
        at android.app.Activity.onResume(Activity.java:1343)
```
```
D/StrictMode: StrictMode policy violation; ~duration=93 ms: android.os.strictmode.DiskReadViolation
        at android.os.StrictMode$AndroidBlockGuardPolicy.onReadFromDisk(StrictMode.java:1504)
        at java.io.UnixFileSystem.checkAccess(UnixFileSystem.java:251)
        at java.io.File.exists(File.java:815)
        at android.app.ContextImpl.ensurePrivateDirExists(ContextImpl.java:605)
        at android.app.ContextImpl.ensurePrivateDirExists(ContextImpl.java:596)
        at android.app.ContextImpl.getDatabasesDir(ContextImpl.java:835)
        at android.app.ContextImpl.getDatabasePath(ContextImpl.java:814)
        at android.content.ContextWrapper.getDatabasePath(ContextWrapper.java:327)
        at android.database.sqlite.SQLiteOpenHelper.getDatabaseLocked(SQLiteOpenHelper.java:352)
        at android.database.sqlite.SQLiteOpenHelper.getWritableDatabase(SQLiteOpenHelper.java:298)
        at com.instabug.library.internal.storage.cache.db.SQLiteDatabaseWrapper.open(SQLiteDatabaseWrapper.java:3)
        at com.instabug.library.internal.storage.cache.db.DatabaseManager.openDatabase(DatabaseManager.java:1)
        at com.instabug.library.internal.storage.cache.user.a.a(UserDbHelper.java:3)
        at com.instabug.library.internal.storage.cache.user.UserCacheManager.insertUser(UserCacheManager.java:2)
        at com.instabug.library.internal.storage.cache.user.UserCacheManager.insertIfNotExists(UserCacheManager.java:3)
        at com.instabug.library.user.b.g(UserManager.java:20)
        at com.instabug.library.b.x(InstabugDelegate.java:2)
        at com.instabug.library.b.y(InstabugDelegate.java:9)
        at com.instabug.library.b.h(InstabugDelegate.java:1)
        at com.instabug.library.b$a.a(InstabugDelegate.java:19)
        at com.instabug.library.b$a.accept(InstabugDelegate.java:1)
        at io.reactivex.internal.observers.LambdaObserver.onNext(LambdaObserver.java:63)
        at io.reactivex.subjects.PublishSubject$PublishDisposable.onNext(PublishSubject.java:308)
        at io.reactivex.subjects.PublishSubject.onNext(PublishSubject.java:228)
        at com.instabug.library.core.eventbus.EventBus.post(EventBus.java:1)
        at com.instabug.library.g.a(SessionManager.java:14)
        at com.instabug.library.g.e(SessionManager.java:3)
        at com.instabug.library.tracking.a.onActivityResumed(InstabugActivityLifecycleListener.java:15)
        at android.app.Application.dispatchActivityResumed(Application.java:239)
```
```
D/StrictMode: StrictMode policy violation; ~duration=92 ms: android.os.strictmode.DiskReadViolation
        at android.os.StrictMode$AndroidBlockGuardPolicy.onReadFromDisk(StrictMode.java:1504)
        at android.database.sqlite.SQLiteConnection.applyBlockGuardPolicy(SQLiteConnection.java:1055)
        at android.database.sqlite.SQLiteConnection.executeForLong(SQLiteConnection.java:609)
        at android.database.sqlite.SQLiteConnection.setPageSize(SQLiteConnection.java:253)
        at android.database.sqlite.SQLiteConnection.open(SQLiteConnection.java:215)
        at android.database.sqlite.SQLiteConnection.open(SQLiteConnection.java:195)
        at android.database.sqlite.SQLiteConnectionPool.openConnectionLocked(SQLiteConnectionPool.java:503)
        at android.database.sqlite.SQLiteConnectionPool.open(SQLiteConnectionPool.java:204)
        at android.database.sqlite.SQLiteConnectionPool.open(SQLiteConnectionPool.java:196)
        at android.database.sqlite.SQLiteDatabase.openInner(SQLiteDatabase.java:880)
        at android.database.sqlite.SQLiteDatabase.open(SQLiteDatabase.java:865)
        at android.database.sqlite.SQLiteDatabase.openDatabase(SQLiteDatabase.java:739)
        at android.database.sqlite.SQLiteDatabase.openDatabase(SQLiteDatabase.java:729)
        at android.database.sqlite.SQLiteOpenHelper.getDatabaseLocked(SQLiteOpenHelper.java:355)
        at android.database.sqlite.SQLiteOpenHelper.getWritableDatabase(SQLiteOpenHelper.java:298)
        at com.instabug.library.internal.storage.cache.db.SQLiteDatabaseWrapper.open(SQLiteDatabaseWrapper.java:3)
        at com.instabug.library.internal.storage.cache.db.DatabaseManager.openDatabase(DatabaseManager.java:1)
        at com.instabug.library.internal.storage.cache.user.a.a(UserDbHelper.java:3)
        at com.instabug.library.internal.storage.cache.user.UserCacheManager.insertUser(UserCacheManager.java:2)
        at com.instabug.library.internal.storage.cache.user.UserCacheManager.insertIfNotExists(UserCacheManager.java:3)
        at com.instabug.library.user.b.g(UserManager.java:20)
        at com.instabug.library.b.x(InstabugDelegate.java:2)
        at com.instabug.library.b.y(InstabugDelegate.java:9)
        at com.instabug.library.b.h(InstabugDelegate.java:1)
        at com.instabug.library.b$a.a(InstabugDelegate.java:19)
        at com.instabug.library.b$a.accept(InstabugDelegate.java:1)
        at io.reactivex.internal.observers.LambdaObserver.onNext(LambdaObserver.java:63)
        at io.reactivex.subjects.PublishSubject$PublishDisposable.onNext(PublishSubject.java:308)
        at io.reactivex.subjects.PublishSubject.onNext(PublishSubject.java:228)
        at com.instabug.library.core.eventbus.EventBus.post(EventBus.java:1)
        at com.instabug.library.g.a(SessionManager.java:14)
        at com.instabug.library.g.e(SessionManager.java:3)
        at com.instabug.library.tracking.a.onActivityResumed(InstabugActivityLifecycleListener.java:15)
        at android.app.Application.dispatchActivityResumed(Application.java:239)
```

tons of them.
Looks like all start-up initialization happens here

**Comment by amrbrkt** [comment]
Thanks @stari4ek for your default report, this was very helpful. 

I've created a snapshot for you contains a fix for these warnings, please note this snapshot is totally safe to use and does not contain any debugging/ non-production code. 

**To use the snapshot:** 
Add these to your app-level `build.gradle `file to grab snapshots:

```
dependencies {
  compile 'com.instabug.library:instabug:4.x.y.z-SNAPSHOT'
}
```
```
repositories {
    maven {
        url "https://oss.sonatype.org/content/repositories/snapshots"
    }
} 
```
Please give it a try and let me know if you still facing these warnings. 
Thanks

**Comment by stari4ek** [comment]
Does not solve the issue at all
```
implementation 'com.instabug.library:instabug:9.0.4.5-SNAPSHOT'
```
```
2020-01-16 12:59:42.951 4592-4592/? D/StrictMode: StrictMode policy violation; ~duration=11 ms: android.os.strictmode.DiskReadViolation
        at android.os.StrictMode$AndroidBlockGuardPolicy.onReadFromDisk(StrictMode.java:1504)
        at java.io.UnixFileSystem.getSpace(UnixFileSystem.java:361)
        at java.io.File.getUsableSpace(File.java:1868)
        at com.instabug.library.util.DeviceStateProvider.getFreeStorage(DeviceStateProvider.java:2)
        at com.instabug.library.tracking.a.onActivityResumed(InstabugActivityLifecycleListener.java:5)
        at android.app.Application.dispatchActivityResumed(Application.java:239)
        at android.app.Activity.onResume(Activity.java:1343)
        at androidx.fragment.app.FragmentActivity.onResume(FragmentActivity.java:456)
        at com.trello.rxlifecycle3.components.support.RxFragmentActivity.onResume(RxFragmentActivity.java:87)
        at by.stari4ek.iptv4atv.tvinput.ui.main.MainMenuActivity.onResume(MainMenuActivity.java:68)
        at android.app.Instrumentation.callActivityOnResume(Instrumentation.java:1412)
        at android.app.Activity.performResume(Activity.java:7292)
        at android.app.ActivityThread.performResumeActivity(ActivityThread.java:3776)
        at android.app.ActivityThread.handleResumeActivity(ActivityThread.java:3816)
        at android.app.servertransaction.ResumeActivityItem.execute(ResumeActivityItem.java:51)
        at android.app.servertransaction.TransactionExecutor.executeLifecycleState(TransactionExecutor.java:145)
        at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:70)
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1808)
        at android.os.Handler.dispatchMessage(Handler.java:106)
        at android.os.Looper.loop(Looper.java:193)
        at android.app.ActivityThread.main(ActivityThread.java:6669)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:493)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:858)
```

**Comment by stari4ek** [comment]
Any updates here, since at least one of case is not fixed with 9.0.4.5-SNAPSHOT

**Comment by stari4ek** [comment]
And another one for each `TRIM_MEMORY_UI_HIDDEN`:
```
D/StrictMode: StrictMode policy violation; ~duration=154 ms: android.os.strictmode.DiskReadViolation
        at android.os.StrictMode$AndroidBlockGuardPolicy.onReadFromDisk(StrictMode.java:1504)
        at android.database.sqlite.SQLiteConnection.applyBlockGuardPolicy(SQLiteConnection.java:1077)
        at android.database.sqlite.SQLiteConnection.executeForLong(SQLiteConnection.java:631)
        at android.database.sqlite.SQLiteConnection.setAutoCheckpointInterval(SQLiteConnection.java:264)
        at android.database.sqlite.SQLiteConnection.open(SQLiteConnection.java:220)
        at android.database.sqlite.SQLiteConnection.open(SQLiteConnection.java:196)
        at android.database.sqlite.SQLiteConnectionPool.openConnectionLocked(SQLiteConnectionPool.java:503)
        at android.database.sqlite.SQLiteConnectionPool.open(SQLiteConnectionPool.java:204)
        at android.database.sqlite.SQLiteConnectionPool.open(SQLiteConnectionPool.java:196)
        at android.database.sqlite.SQLiteDatabase.openInner(SQLiteDatabase.java:887)
        at android.database.sqlite.SQLiteDatabase.open(SQLiteDatabase.java:865)
        at android.database.sqlite.SQLiteDatabase.openDatabase(SQLiteDatabase.java:739)
        at android.database.sqlite.SQLiteDatabase.openDatabase(SQLiteDatabase.java:729)
        at android.database.sqlite.SQLiteOpenHelper.getDatabaseLocked(SQLiteOpenHelper.java:355)
        at android.database.sqlite.SQLiteOpenHelper.getWritableDatabase(SQLiteOpenHelper.java:298)
        at com.instabug.library.internal.storage.cache.db.SQLiteDatabaseWrapper.open(SQLiteDatabaseWrapper.java:3)
        at com.instabug.library.internal.storage.cache.db.DatabaseManager.openDatabase(DatabaseManager.java:1)
        at com.instabug.library.internal.storage.cache.user.a.a(UserDbHelper.java:38)
        at com.instabug.library.internal.storage.cache.user.UserCacheManager.insertIfNotExists(UserCacheManager.java:2)
        at com.instabug.library.user.b.g(UserManager.java:20)
        at com.instabug.library.internal.storage.cache.db.userAttribute.UserAttributeCacheManager.retrieveAll(UserAttributeCacheManager.java:1)
        at com.instabug.library.internal.storage.cache.db.userAttribute.UserAttributesDbHelper.getAll(UserAttributesDbHelper.java:1)
        at com.instabug.library.g.g(SessionManager.java:1)
        at com.instabug.library.g.a(SessionManager.java:6)
        at com.instabug.library.g.f(SessionManager.java:2)
        at com.instabug.library.g.m(SessionManager.java:3)
        at com.instabug.library.g.b(SessionManager.java:4)
        at com.instabug.library.tracking.a.onTrimMemory(InstabugActivityLifecycleListener.java:7)
        at android.app.Application.onTrimMemory(Application.java:149)
```

**Comment by Muhammadyoussef** [comment]
Hi @stari4ek ,

A quick update: We are working hard on this issue and making sure no similar issues will arise in the future. The fix is now in our testing pipeline and expected to be released within the next two weeks.

**Comment by stari4ek** [comment]
"better now than never".

BTW, dropping Instabug from the project made `-600ms` from median app start time (it's a lot!). And `-1.75s` on 95% percentile.

**Comment by carstenhag** [comment]
FYI, has been fixed with 9.1.6 apparently. https://github.com/Instabug/Instabug-Android/releases/tag/9.1.6

**Comment by MahmoudAmin39** [comment]
Hi @stari4ek ,
This issue should be fixed on our latest Sdk version `9.1.6`.
Please upgrade to the latest version and feedback us if you are still facing any issue.
I will be closing this issue for now. Feel free to re-open it.

Thank you.
